### PR TITLE
Stream Enrich: add the possibility to override the DynamoDB endpoint (closes #3942)

### LINKED ISSUE #3942: Stream Enrich: add the possibility to override the DynamoDB endpoint
**Project**: Stream Enrich

**Version**: r111

**Expected behavior**: Be able to override Dyanmodb endpoint to the localhost one

**Actual behavior**: This feature is missing

**Steps to reproduce**: Using Localstack instead of AWS

 1. We need to override the Endpoints for Dynamodb, because KCL keeps the checkpoints there. 
 2. snowplow/snowplow#3344  has address it different by the region selection. but we need to add the whole endpoint URL for it.
3. I propose that we add a setting similar to kinesisEndpoint in the settings file .

### PR DESCRIPTION
**Pull request by szareiangm**
- We have added the optional endpoint to Dynamodb to make it work with Localstack.
- Also, Cloudwatch metrics was not behaving as expected with Localstack so we needed to use the NullFactory to disable it.

### CODE REVIEW

**Comment by szareiangm** [comment]
Not sure what has failed in the build. can someone please restart it?

**Comment by chuwy** [comment]
@szareiangm done.

File: 3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/Amodel.scala [hunk 1]
```diff
@@ -64,6 +66,17 @@ object model {
...
       case _                 => s"https://kinesis.$region.amazonaws.com"
     })
+
+
+    val dynamodbEndpoint = dynamodbCustomEndpoint.getOrElse(region match {
+      case cn @ "cn-north-1" => s"https://dynamodb.$cn.amazonaws.com.cn"
+      case _                 => s"https://dynamodb.$region.amazonaws.com"
+    })
+
+    val iscloudwatchDisabled = disableCloudWatch.getOrElse(region match {
```
Comment (benjben) [inline 3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/Amodel.scala]: You don't need this.

File: 3-enrich/stream-enrich/kinesis/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/KinesisSource.scala [hunk 2]
```diff
@@ -140,11 +142,18 @@ class KinesisSource private (
     log.info(s"Processing raw input stream: ${config.in.raw}")
 
     val rawEventProcessorFactory = new RawEventProcessorFactory()
-    val worker = new Worker(
-      rawEventProcessorFactory,
-      kinesisClientLibConfiguration
-    )
-
+    val worker = if (kinesisConfig.disableCloudWatch.isDefined && kinesisConfig.iscloudwatchDisabled == true){
```
Comment (benjben) [inline 3-enrich/stream-enrich/kinesis/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/KinesisSource.scala]: You can directly use `kinesisConfig.disableCloudWatch` :
```
val worker = kinesisConfig.disableCloudWatch match {
  case Some(true) =>
    new Worker(
      rawEventProcessorFactory,
      kinesisClientLibConfiguration,
      new NullMetricsFactory()
    )
  case _ =>
    new Worker(
      rawEventProcessorFactory,
      kinesisClientLibConfiguration
    )
}
```

File: 3-enrich/stream-enrich/examples/config.hocon.sample [hunk 3]
```diff
@@ -69,6 +69,13 @@ enrich {
       # customEndpoint = {{kinesisEndpoint}}
       # customEndpoint = ${?ENRICH_STREAMS_SOURCE_SINK_CUSTOM_ENDPOINT}
 
+      ## Optional endpoint url configuration to override aws dyanomdb endpoints,
```
Comment (benjben) [inline 3-enrich/stream-enrich/examples/config.hocon.sample]: Can you please precise that this is for Kinesis checkpoints.

File: 3-enrich/stream-enrich/examples/config.hocon.sample [hunk 4]
```diff
@@ -69,6 +69,13 @@ enrich {
...
+      ## this can be used to specify local endpoints when using localstack
+      # dynamodbCustomEndpoint = "http://localhost:4569"
+
+      # Optional override to disable cloudwatch; set to true to disable
+      # disableCloudWatch = false
```
Comment (benjben) [inline 3-enrich/stream-enrich/examples/config.hocon.sample]: =>
```
# Optional override to disable cloudwatch
# disableCloudWatch = true
```

File: 3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/Amodel.scala (hunk 1 above)
Comment (szareiangm) [reply 3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/Amodel.scala]: removed.

File: 3-enrich/stream-enrich/examples/config.hocon.sample (hunk 3 above)
Comment (szareiangm) [reply 3-enrich/stream-enrich/examples/config.hocon.sample]: done

File: 3-enrich/stream-enrich/examples/config.hocon.sample (hunk 4 above)
Comment (szareiangm) [reply 3-enrich/stream-enrich/examples/config.hocon.sample]: done

File: 3-enrich/stream-enrich/kinesis/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/KinesisSource.scala (hunk 2 above)
Comment (szareiangm) [reply 3-enrich/stream-enrich/kinesis/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/KinesisSource.scala]: done

**Comment by szareiangm** [comment]
I have addressed all of the comments. This works is part of the project that our Summer co-op student, Saud Naumani, has done. I just submitted it and addressed the comments because he went back to the university.

**Comment by szareiangm** [comment]
also, you might want to run `scalafmt` on stream-enrich for the next release preparation commit. I could run it but the files are too many for the purpose of this ticket.

**Comment by benjben** [comment]
@szareiangm could you please update this PR so that it can be merged into `release/r119` ?

**Comment by szareiangm** [comment]
Hi, sure. I will give it a shot until Monday.

**Comment by benjben** [comment]
Great thanks!

**Comment by benjben** [comment]
Hi @szareiangm ,
As we are going to release R119 soon and we'd like this feature to be in it, I reapplied the changes manually on another branch and used your name and email as author : https://github.com/snowplow/snowplow/pull/4327
Thanks for the contribution!

**Pull request closed without merging**